cpp/src/arrow/flight/transport/ucx/ucx_server.cc

```diff
@@ -0,0 +1,646 @@
+ public:
+  explicit PutServerStream(UcpCallDriver* driver)
+      : UcxServerStream(driver), finished_(false) {}
+
+  bool ReadData(internal::FlightData* data) override {
+    if (finished_) return false;
+
+    bool success = true;
+    auto status = ReadImpl(data).Value(&success);
+
+    if (!status.ok() || !success) {
+      finished_ = true;
+      if (!status.ok()) {
+        FLIGHT_LOG_PEER(WARNING, peer_) << "I/O error in DoPut: " << status.ToString();
+        return false;
+      }
+    }
+    return success;
+  }
+
+  Status WritePutMetadata(const Buffer& payload) override {
+    if (finished_) return Status::OK();
+    // Send synchronously (we don't control payload lifetime)
+    return driver_->SendFrame(FrameType::kBuffer, payload.data(), payload.size());
+  }
+
+ private:
+  ::arrow::Result<bool> ReadImpl(internal::FlightData* data) {
+    ARROW_ASSIGN_OR_RAISE(auto frame, driver_->ReadNextFrame());
+    if (frame->type == FrameType::kHeaders) {
+      // Trailers, client is done writing
+      return false;
+    }
+    RETURN_NOT_OK(driver_->ExpectFrameType(*frame, FrameType::kPayloadHeader));
+    PayloadHeaderFrame payload_header(std::move(frame->buffer));
+    RETURN_NOT_OK(payload_header.ToFlightData(data));
+
+    if (data->metadata) {
+      ARROW_ASSIGN_OR_RAISE(auto message, ipc::Message::Open(data->metadata, nullptr));
+
+      if (ipc::Message::HasBody(message->type())) {
+        ARROW_ASSIGN_OR_RAISE(frame, driver_->ReadNextFrame());
+        RETURN_NOT_OK(driver_->ExpectFrameType(*frame, FrameType::kPayloadBody));
+        data->body = std::move(frame->buffer);
+      }
+    }
+    return true;
+  }
+
+  bool finished_;
+};
+
+class ExchangeServerStream : public PutServerStream {
+ public:
+  using PutServerStream::PutServerStream;
+
+  arrow::Result<bool> WriteData(const FlightPayload& payload) override {
+    if (writes_done_) return false;
+    // Don't use backpressure - the application may expect synchronous
+    // behavior (write a message, read the client response)
+    Future<> pending_send = driver_->SendFlightPayload(payload);
+    while (!pending_send.is_finished()) {
+      driver_->MakeProgress();
+    }
+    RETURN_NOT_OK(pending_send.status());
+    return true;
+  }
+  Status WritePutMetadata(const Buffer& payload) override {
+    return Status::NotImplemented("Not supported on this stream");
+  }
+};
+
+class UcxServerImpl : public arrow::flight::internal::ServerTransport {
+ public:
+  using arrow::flight::internal::ServerTransport::ServerTransport;
+
+  virtual ~UcxServerImpl() {
+    if (listening_.load()) {
+      auto st = Shutdown();
+      if (!st.ok()) {
+        ARROW_LOG(WARNING) << "Server did not shut down properly: " << st.ToString();
+      }
+    }
+  }
+
+  Status Init(const FlightServerOptions& options,
+              const arrow::internal::Uri& uri) override {
+    const auto num_threads = std::max<uint32_t>(8, std::thread::hardware_concurrency());
```

Comment:
Will a large `num_threads` cause trouble? Some Arm servers in production even have 200+ cores.

Reply:
It should not, this is a max # of threads (I've renamed it to make that clearer) and the pool will only create as many threads as needed. One RPC is handled per thread so this will let us handle up to (numcores) concurrent requests (the rest will just be queued).